File: answers/leetcode/Best Time to Buy and Sell Stock with Cooldown/Best Time to Buy and Sell Stock with Cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int buy = INT_MIN;
        int pre_buy = 0;
        int sell = 0;
        int pre_sell = 0;
        for (size_t i = 0; i < prices.size(); ++ i) {
            pre_buy = buy;
            buy = max(buy, pre_sell - prices[i]);
            pre_sell = sell;
            sell = max(sell, pre_buy + prices[i]);
        }
        return sell;
    }
};

File: answers/leetcode/Best Time to Buy and Sell Stock with Cooldown/Best Time to Buy and Sell Stock with Cooldown.java
class Solution {
    public int maxProfit(int[] prices) {
        int sell0 = 0;
        int buy0 = 0;
        int sell = 0;
        int buy = Integer.MIN_VALUE;
        for (int price : prices) {
            buy0 = buy;
            buy = Math.max(buy0, sell0 - price);
            sell0 = sell;
            sell = Math.max(sell0, buy0 + price);
        }
        return sell;
    }
}